File: src/main/java/net/tecdroid/subsystems/wrist/WristConfiguration.java
package net.tecdroid.subsystems.wrist;

import static edu.wpi.first.units.Units.*;
import static net.tecdroid.subsystems.wrist.WristController.*;
import static net.tecdroid.util.RotationalDirection.Clockwise;
import static com.ctre.phoenix6.signals.NeutralModeValue.*;

import com.ctre.phoenix6.signals.NeutralModeValue;
import edu.wpi.first.units.measure.Angle;
import edu.wpi.first.units.measure.AngularVelocity;
import edu.wpi.first.units.measure.Current;
import edu.wpi.first.units.measure.Time;
import net.tecdroid.util.*;


public class WristConfiguration {

    private static class DeviceIDs {
        private final DigitId wristDigit = new DigitId(6);
        private final DigitId throughborePort = new DigitId(0);
        private final DigitId wristMotorDigit = new DigitId(1);
        final DeviceIdentifiers wristDevicesIDs = new DeviceIdentifiers(
                IdentifiersKt.joinDigits(wristDigit, wristMotorDigit),
                throughborePort
        );
    }

    private static class Devices {
        private final MotorProperties wristMotorProperties = Motors.INSTANCE.getKrakenX60();
        final DeviceProperties wristDeviceProperties = new DeviceProperties(wristMotorProperties);
    }

    private static class Limits {
        private final Current motorsCurrentLimit = Amps.of(30.0);
        private final Angle minimumAngleLimit = Degrees.of(-30.0); // TODO: GET REAL MINIMUM. ARBITRARY VALUE
        private final Angle maximumAngleLimit = Degrees.of(45.0); // TODO: GET REAL MAXIMUM. ARBITRARY VALUE
        private final AngularVelocity maximumAngularVelocity = RotationsPerSecond.of(33.3333); // A third of motor's capacity
        final DeviceLimits wristDeviceLimits = new DeviceLimits(
                motorsCurrentLimit, minimumAngleLimit, maximumAngleLimit, maximumAngularVelocity);
    }

    private static class Conventions {
        private final RotationalDirection wristPositiveDirection = Clockwise;
        final DeviceConventions wristDeviceConventions = new DeviceConventions(wristPositiveDirection);
    }

    private static class Structure {
        private final GearRatio wristGR = new GearRatio(89.2857, 1.0, 0);
        private final Angle encoderOffset = Degrees.of(45.0); // TODO: GET REAL OFFSET. ARBITRARY VALUE
        final PhysicalDescription wristPhysicalDescription = new PhysicalDescription(wristGR, encoderOffset);
    }

    private static class Control {
        // TODO: Determine all PIDF, SVAG coefficients
        private final PidfCoefficients motorPIDF = new PidfCoefficients(0.0, 0.0, 0.0, 0.0);
        private final SvagGains motorSVAG = new SvagGains(0.0, 0.0, 0.0, 0.0);
        private final NeutralModeValue motorNeutralMode = Brake;
        private final Time motorsRampRate = Seconds.of(0.1);
        final ControlConstants wristControlConstants = new ControlConstants(
                motorPIDF, motorSVAG, motorNeutralMode, motorsRampRate);
    }

    final Config wristConfig = new Config(
            new DeviceIDs().wristDevicesIDs, new Devices().wristDeviceProperties, new Limits().wristDeviceLimits,
            new Conventions().wristDeviceConventions, new Structure().wristPhysicalDescription, new Control().wristControlConstants
    );
}

File: src/main/java/net/tecdroid/subsystems/wrist/WristConfiguration.kt
package net.tecdroid.subsystems.wrist

import edu.wpi.first.units.AngleUnit
import edu.wpi.first.units.Units.Second
import edu.wpi.first.units.measure.Angle
import edu.wpi.first.units.measure.Current
import net.tecdroid.mechanical.Reduction
import net.tecdroid.util.*
import net.tecdroid.util.RotationalDirection.Clockwise
import net.tecdroid.util.units.amps
import net.tecdroid.util.units.rotations
import net.tecdroid.util.units.seconds
import net.tecdroid.safety.SubsystemLimits

data class WristConfig(
    val motorControllerId: NumericId,
    val motorDirection: RotationalDirection,
    val motorCurrentLimit: Current,

    val absoluteEncoderPort: NumericId,
    val absoluteEncoderIsInverted: Boolean,
    val absoluteEncoderOffset: Angle,

    val reduction: Reduction,
    val limits: SubsystemLimits<AngleUnit>,
    val controlGains: ControlGains,
    val motionTargets: AngularMotionTargets,
)

val wristConfig = WristConfig(
    motorControllerId = NumericId(61),
    motorDirection = Clockwise,
    motorCurrentLimit = 30.0.amps,

    absoluteEncoderPort = NumericId(2),
    absoluteEncoderIsInverted = false,
    absoluteEncoderOffset = (0.1501).rotations,

    reduction = Reduction(214.285714),

    limits = SubsystemLimits(
        absoluteMinimum = 0.0.rotations,
        relativeMinimum = 0.021.rotations,
        relativeMaximum = 0.3704.rotations,
        absoluteMaximum = 0.3748.rotations,
    ),

    controlGains = ControlGains(
        p = 0.1,
        s = 0.1191,
        v = 0.11037,
        a = 0.0019686,
        g = 0.0012195
    ),

    motionTargets = AngularMotionTargets(
        cruiseVelocity = 0.5.rotations.per(Second),
        accelerationTimePeriod = 0.5.seconds,
        jerkTimePeriod = 0.1.seconds
    )
)